go-controller/pkg/ovn/pods.go

```diff
@@ -374,30 +376,44 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) (err error) {
 			return fmt.Errorf("unable to create LSPSetDynamicAddresses command for port: %s", portName)
 		}
 		cmds = append(cmds, cmd)
-	} else {
+
+		// ensure we have/can allocate the IP in the annotation
+		if err = oc.lsManager.AllocateIPs(logicalSwitch, podIfAddrs); err != nil && err != ipallocator.ErrAllocated {
+			// this should also be treated as an allocation for purposes of error handling
+			releaseIPs = true
+			klog.Warningf("Unable to allocate IPs for already annotated pod: %s, IPs: %s, error: %v", pod.Name,
+				util.JoinIPNetIPs(podIfAddrs, " "), err)
+		} else {
+			needsIP = false
+		}
+	}
+
+	if needsIP {
+		// try to get the IP from existing port in OVN first
 		podMac, podIfAddrs, err = oc.getPortAddresses(logicalSwitch, portName)
 		if err != nil {
 			return fmt.Errorf("failed to get pod addresses for pod %s on node: %s, err: %v",
 				portName, logicalSwitch, err)
 		}
-		if podMac == nil || podIfAddrs == nil {
+		needsNewAssignment := false
+		// ensure we have/can allocate the IP found in OVN
+		if len(podIfAddrs) == 0 {
+			needsNewAssignment = true
+		} else if err = oc.lsManager.AllocateIPs(logicalSwitch, podIfAddrs); err != nil && err != ipallocator.ErrAllocated {
+			klog.Warningf("Unable to allocate IPs found on existing OVN port: %s, for pod %s on node: %s"+
+				" error: %v", util.JoinIPNetIPs(podIfAddrs, " "), portName, logicalSwitch, err)
+
+			needsNewAssignment = true
```

Comment:
We cannot change allocation on a running pod. If the pod has an annotation it must be honored. We can't change it in ovnkube master because that races against the node which might be setting up this pod right now, and it won't care that we changed the annotation, it'll just use the old one and we'll end up with an annotation that's inconsistent with the actual pod IP.

Reply:
I think this comment is in the wrong place, should be on line 384 or so. If we cannot allocate the IP that is in the annotation, then it sounds like we cannot assign a new one, and treat that as an error instead. I'll make that change.